MonkeyWrench.Web.WebService/Scheduler/Scheduler.cs

```diff
@@ -400,11 +395,43 @@ private static void AddWork (DB db, List<DBHost> hosts, List<DBLane> lanes, List
 								}
 							}
 
-							sql.AppendFormat ("UPDATE RevisionWork SET state = {0} WHERE id = {1} AND state = 10;", (int) (has_dependencies ? DBState.DependencyNotFulfilled : DBState.NotDone), revisionwork.id);
+							sql.AppendFormat ("UPDATE RevisionWork SET state = 11 WHERE id = {0} AND state = 10;", revisionwork.id);
+							if (!editedHostLanes.Contains (hostlane)) editedHostLanes.Add (hostlane);
+						}
+					}
+				}
 
+				foreach (var editedHostLane in editedHostLanes) {
+					int mostRecent = -1;
+					int mostRecentState = -1;
+					using (IDbCommand cmd = db.CreateCommand ()) {
+						cmd.CommandText = @"
+SELECT RevisionWork.id, RevisionWork.state
+FROM RevisionWork
+INNER JOIN Revision ON Revision.id = RevisionWork.revision_id
+WHERE
+     RevisionWork.host_id = @host_id
+ AND RevisionWork.lane_id = @lane_id
+ORDER BY Revision.Date DESC
+LIMIT 1
+";
+						DB.CreateParameter (cmd, "host_id", editedHostLane.host_id);

```

Comment:
The first query in your suggestion takes about 7s to run. Each of the SELECTs in the format I had written takes about 17ms to run. Though we will be hitting the DB more often, I think it's worth that overhead.